### PR TITLE
feat: Add web service

### LINKED ISSUE #4: Web service
- [ ] Fastify app
- [ ] Listen on port
- [ ] `POST /:subnet/measurement` updates rsr stats
- [ ] `GET /:subnet` returns rsr stats
- [ ] Basic test suite

Blocked by #3
Blocks #5

### PR DESCRIPTION
**Pull request by pyropy**
## Description

This PR introduces `measurements` table two subnet endpoints. Table `measurements` is ment to save RSR stats for different subnets.

New endpoint `/:subnet` returns RSR score for subnet and endpoint `/:subnet/measurements` collects measurements submitted by checker nodes.

## Change-log

- Adds measurements table that collects measurements RSR for different subnets
- Adds subnet routes for reading and updating subnet RSR
- Adds basic tests for new routes

Closes #4

### CODE REVIEW

File: lib/request-helpers.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ */
+export const subnetPreHandlerHook = async (request, reply) => {
+  // @ts-ignore TODO: Fix
+  const { subnet } = request.params
+  if (!subnet || typeof subnet !== 'string') {
+    reply.code(400).send({ error: 'Invalid subnet parameter' })
+    return
+  }
+
+  if (!isValidSubnet(subnet.toLowerCase())) {
```
Comment (bajtos) [inline lib/request-helpers.js:14]: Is it necessary to convert `subnet` to lower case? Shouldn't we require the clients to use the canonical (lower case) form?

Your current solution follows the [Robustness Principle](https://en.wikipedia.org/wiki/Robustness_principle) - _be conservative in what you do, be liberal in what you accept from others_. It is important when you don't control the other party. However, we control both the server accepting the requests and the client making the requests, so I think we can afford to be strict here.

Anyhow, I am fine with keeping `subnet.toLowerCase()` if you prefer, but please add a test to show how non-lower-case values are handled.

File: lib/subnet-routes.js [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+          [request.body ? 1 : 0, request.params.subnet]
+        )
+      } finally {
+        client.release()
+      }
+    }
+  )
+
+  app.get(
+    '/:subnet',
```
Comment (bajtos) [inline lib/subnet-routes.js:39]: Let's make the path more descriptive please. `GET /walrus` does not communicate what is the expected response.

Looking at https://github.com/CheckerNetwork/spark-stats/blob/main/README.md, I propose the following endpoint name:

```suggestion
    '/:subnet/retrieval-success-rate',
```

File: lib/subnet-routes.js [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    },
+    /**
+     * @param {RequestWithSubnet} request
+     * @param {FastifyReply} reply
+     */
+    async (request, reply) => {
+      const client = await app.pg.connect()
+      try {
+        await client.query(
+          'UPDATE measurements SET total = total + 1, successful = successful + $1 WHERE subnet = $2',
```
Comment (bajtos) [inline lib/subnet-routes.js:29]: TL;DR: I think it's important to include `day` in the data schema from the beginning.

The current design will give us only "all-time" stats, effectively averaging daily RSR over the entire time we collect measurements. I find that problematic.

Let's say the network has low RSR for many weeks and then suddenly they fix retrievability and RSR jumps up. It will take a long time until the average value goes up to show improvement in RSR.

For example:

- On days 03-01 to 03-31, we measure RSR = 10%.
- On days 04-01 to 04-06, we measure RSR = 90%.
- On 04-07, the "all-time" RSR is 24.7%, while the "7-day average" is 90%.

I appreciate this may be viewed as feature creep or scope inflation compared to the original plan. I am happy to discuss more.

File: lib/request-helpers.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    reply.code(404).send()
+    console.log('Invalid subnet parameter')
+  }
+}
+
+/**
+ * Validate if the given subnet is either 'arweave' or 'walrus'
+ * @param {string} subnet
+ * @returns {boolean}
```
Comment (bajtos) [inline lib/request-helpers.js:23]: If you write something like the following (I don't remember the exact TS syntax), then TypeScript will be able to understand that after calling `isValidSubnet`, the `subnet` variable is no longer "arbitrary string" but must be `"walrus" | "arweave"`.

```suggestion
 * @returns {subnet is Subnet}
```

File: lib/request-helpers.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+export const subnetPreHandlerHook = async (request, reply) => {
+  // @ts-ignore TODO: Fix
+  const { subnet } = request.params
+  if (!subnet || typeof subnet !== 'string') {
+    reply.code(400).send({ error: 'Invalid subnet parameter' })
+    return
+  }
+
+  if (!isValidSubnet(subnet.toLowerCase())) {
+    reply.code(404).send()
```
Comment (bajtos) [inline lib/request-helpers.js:15]: In my experience with cid.contact that returns 404 for unknown CIDs, it's difficult to tell whether you have a typo in the requested path or whether you used invalid value for one of the path parameters.

For example, both `POST /walrus/measurment` and `POST /filecoin/measurement` return the same code 404. The first one contains a typo in the last path segment, the second is using an unsupported subnet name.

Let's return the status code 400 instead. You already return 400 when the subnet is not provided.

```suggestion
    reply.code(400).send()
```

File: test/subnet-routes.test.js [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+  describe('/:subnet/measurement', () => {
+    it('submit successful measurement', async () => {
+      const subnet = 'walrus'
+      await withSubnetMeasurements(pgPool, subnet, 0, 0)
+      const res = await fetch(new URL(`/${subnet}/measurement`, baseUrl), {
+        method: 'POST',
+        headers: { 'Content-Type': 'application/json' },
+        body: JSON.stringify(true)
+      })
```
Comment (bajtos) [inline test/subnet-routes.test.js:54]: Is it worth extracting a helper for POSTing JSON?

```js
// usage
const res = await postJson(new URL(`/${subnet}/measurement`, baseUrl), true)

// implementation
function postJson(url, body) {
  return fetch(new URL(`/${subnet}/measurement`, baseUrl), {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body)
  })
}  
```

File: lib/subnet-routes.js [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+  app.addHook('preHandler', subnetPreHandlerHook)
+
+  app.post(
+    '/:subnet/measurement',
+    {
+      schema: {
+        body: {
+          type: 'boolean'
+        }
+      }
```
Comment (bajtos) [inline lib/subnet-routes.js:19]: "nake" booleans are usually problematic, because it's not clear what the true/false value refers to. Also, if the request body is not an object but just a boolean value, then it's not open to extension - adding more measurement fields becomes a breaking change.

I know we want to keep this as simple as possible, and we don't have any immediate needs for adding more fields. ([YAGNI](https://en.wikipedia.org/wiki/You_aren't_gonna_need_it)).

However, considering how small the change is, and how much effort it will save us in case we need to send more than one measurement field, I feel it's worth using an extensible schema from the start.

```suggestion
      schema: {
        body: {
          type: 'object',
          properties: {
            retrievalSucceeded: {
              type: 'boolean'
            }
          },
          required: ['retrievalSucceeded']
        }
      }
```

File: lib/subnet-routes.js [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+                type: 'string',
+                format: 'int64',
+                pattern: '^[0-9]*$'
+              },
+              successful: {
+                type: 'string',
+                format: 'int64',
+                pattern: '^[0-9]*$'
+              }
+            }
```
Comment (bajtos) [inline lib/subnet-routes.js:56]: In JSON Schema, all properties are optional by default. That's great for extensibility/forward-compatibility, but it makes it more difficult to consume such data in strictly typed languages - clients have to check whether `total` and `successful` were provided by the server. In our case, we want the server to always provide both properties, so let's describe that in the schema too.

```suggestion
            },
            required: ['total', 'successful']
```

**Review by bajtos — CHANGES_REQUESTED** [review]
Great start!

File: lib/request-helpers.js (hunk 4 above)
Comment (bajtos) [reply lib/request-helpers.js:23]: You can learn more about user-defined type-guards here: https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards

Comment (pyropy) [reply lib/request-helpers.js:23]: Amazing! Didn't know about this feature. Thanks for sharing! 🙌🏻

File: lib/subnet-routes.js (hunk 8 above)
Comment (pyropy) [reply lib/subnet-routes.js:56]: I wasn't aware of that, thank you for pointing it out 🙏🏻

File: test/subnet-routes.test.js (hunk 6 above)
Comment (pyropy) [reply test/subnet-routes.test.js:54]: Good suggestion, it does repeat few times and I wouldn't mind extracting it.

File: lib/request-helpers.js (hunk 1 above)
Comment (pyropy) [reply lib/request-helpers.js:14]: I just found out that we could actually do this validation in schema using regex. It can be done just by passing params to schema definition:

```javascript
app.post(
'/:subnet/measurement',
schema: {
  params: { type: 'object',
    additionalProperties: false,
    properties: {
      subnet: {
        type: 'string',
        pattern: '^(walrus|arweave)$'
      }
    },
    required: ['subnet']
  },
  body: { ... }
},
async (req, reply) => { ... })
```

File: lib/subnet-routes.js (hunk 3 above)
Comment (pyropy) [reply lib/subnet-routes.js:45]: I think it's a fair point that you've raised this up. 

We'll probably want to have measurement stats for given period anyway at some point, so I think it's much better to do it now rather than later.

Comment (pyropy) [reply lib/subnet-routes.js]: @bajtos I have added the new column https://github.com/CheckerNetwork/simple-subnet-api/pull/12/commits/1ec0aa8674044c691180c6c3d88dc491c5420d57 . Apart from that I have also included filtering RSR scores by time range (by default we return scores for today like `spark-stats` does).

File: lib/request-helpers.js (hunk 1 above)
Comment (bajtos) [reply lib/request-helpers.js]: >I just found out that we could actually do this validation in schema using regex. It can be done just by passing params to schema definition

I love it! ❤️

**Review by bajtos — COMMENTED** [review]
The new version looks much better, great progress! 👏🏻 

I have a few more comments to discuss, PTAL 👇🏻

File: lib/subnet-routes.js [hunk 9]
```diff
@@ -36,41 +57,66 @@ export const subnetRoutes = (app) => {
...
-              }
+              required: ['total', 'successful']
             }
           }
         }
       }
     },
     /**
-     * @param {RequestWithSubnet} request
+     * @param {RequestWithSubnet<any, {from: string; to: string}>} request
```
[the remark this answers is not in the document]
Comment (pyropy) [reply lib/subnet-routes.js:105]: Setting it to `{}` sounds ok to me 👍🏻

File: test/subnet-routes.test.js [hunk 10]
```diff
@@ -44,70 +46,112 @@ describe('Subnet routes', () => {
...
-      const res = await fetch(new URL(`/${unknownSubnet}/measurement`, baseUrl), {
-        method: 'POST',
-        headers: { 'Content-Type': 'application/json' },
-        body: JSON.stringify(false)
-      })
-
-      await assertResponseStatus(res, 404)
+      await withSubnetMeasurements({ pgPool, day, subnet, total: 0, successful: 0 })
+      // @ts-ignore - intentional error
+      const res = await postMeasurement(baseUrl, unknownSubnet, true)
```
[the remark this answers is not in the document]
Comment (pyropy) [reply test/subnet-routes.test.js:107]: Thanks for the tips 🫶

File: lib/subnet-routes.js [hunk 11]
```diff
@@ -36,41 +57,66 @@ export const subnetRoutes = (app) => {
...
      * @param {FastifyReply} reply
      */
     async (request, reply) => {
       const client = await app.pg.connect()
+      const from = request.query.from || today()
+      const to = request.query.to || today()
```
[the remark this answers is not in the document]
Comment (pyropy) [reply lib/subnet-routes.js:111]: I have tried to add tests for it and in case we pass down the empty value to query string it won't pass the validation. It does however allow you not to pass one or both query strings and in that case it sets the default value -- which is an expected behavior.

File: test/test-helpers.js [hunk 12]
```diff
@@ -15,20 +15,35 @@ export const assertResponseStatus = async (res, status) => {
...
+ * @param {import('../lib/typings.js').Subnet} subnet
+ * @param {boolean} retrievalSucceeded
+ */
+export const postMeasurement = (baseUrl, subnet, retrievalSucceeded) => {
+  return fetch(new URL(`/${subnet}/measurement`, baseUrl), {
+    method: 'POST',
+    headers: { 'Content-Type': 'application/json' },
+    body: JSON.stringify({ retrievalSucceeded })
+  })
+}
```
[the remark this answers is not in the document]
Comment (pyropy) [reply test/test-helpers.js:49]: I don't see some obvious improvements to be honest but I'm going to accept your suggestion.

File: lib/subnet-routes.js [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+    async (request, reply) => {
+      const client = await app.pg.connect()
+      const from = request.query.from || today()
+      const to = request.query.to || today()
+      try {
+        const { rows } = await client.query(
+          `SELECT day, total, successful FROM daily_measurements
+           WHERE subnet = $1 AND day >= $2 AND day <= $3
+           GROUP BY subnet, day ORDER BY day`,
+          [request.params.subnet, from, to]
```
Comment (NikolasHaimerl) [inline lib/subnet-routes.js:117]: We are passing arbitrary strings from a user to this query. Should we check that the strings are dates first?

Comment (pyropy) [reply lib/subnet-routes.js:117]: Validation of the inputs `from` and `to` is done at schema level. Validated values are of type string but they match the date format. 

Here's the schema definition for query strings:
```javascript
{
   ...
  schema: {
    params: subnetRoutesDefaultParamsSchema,
    querystring: {
      type: 'object',
      properties: {
        from: {
          type: 'string',
          format: 'date'
        },
        to: {
          type: 'string',
          format: 'date'
        }
      }
    }
}
 ```

**Review by NikolasHaimerl — APPROVED** [review]

**Review by bajtos — APPROVED** [review]
:shipit:
